linchpin/provision/roles/libvirt/tasks/provision_libvirt_node.yml

```diff
@@ -331,24 +334,47 @@
     loop_var: definition
   ignore_errors: no
 
-- name: "mac_and_ip | extract mac address"
-  shell: >
-    virsh -c {{ node[1] }} dumpxml {{ node[0] }}_{{ node[2] }}
-    | grep 'mac address'
-    | cut -f 2 -d "'"
+- name: "dump node data"
+  virt:
+    name: "{{ data[0] }}_{{ data[2] }}"
+    command: get_xml
+    uri: "{{ data[1] }}"
   with_nested:
     - ["{{ libvirt_resource_name }}"]
     - ["{{ res_def['uri'] | default('qemu:///system') }}"]
     - "{{ res_count }}"
+  loop_control:
+    loop_var: data
+  when: not async
+  register: node_data
+
+- xml:
+    xmlstring: "{{ node_data.results[net.1].get_xml }}"
+    xpath: "//interface[@type='network']/source[@network='{{net.0.name}}']/../mac"
+    content: attribute
+  register: mac_addresses
+  ignore_errors: yes
+  loop_control:
+    loop_var: net
+  loop: "{{ res_def['networks']|product(res_count)|list }}"
+  when:
+    - res_def['networks'] is defined
+    - net.dhcp is not defined or net.dhcp == true
+
+- set_fact:
+    macs: []
+
+- set_fact:
+    macs: "{{ macs + [mac_addresses.results[node.1].matches[node.0].mac.address] }}"
```

Comment:
"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'matches'\n\nThe error appears to have been in '/usr/lib/python2.7/site-packages/linchpin/provision/roles/libvirt/tasks/provision_libvirt_node.yml': line 367,

Reply:
Right but its actually [failing a bit earlier](https://jenkins-continuous-infra.apps.ci.centos.org/blue/organizations/jenkins/ci-linchpin/detail/ci-linchpin/1130/pipeline/#step-199-log-479) -
`"Xpath //interface[@type='network']/source[@network='default']/../mac does not reference a node!"` which causes `mac_addresses.results[node.1].matches` to be empty array